### DISCUSSION TITLE
GET with Body is necessary on some APIs

### DISCUSSION THREAD

**Discussion by mauricioszabo** [node-fetch/node-fetch#1524] [Ideas]
Referencing this issue: https://github.com/node-fetch/node-fetch/issues/1377

Unfortunately, there _are_ APIs that need GET with body: Shortcut is one, Elastic is another. There may be also private APIs that need this behavior. There's also no mention on the RFC that it's is, indeed, forbidden: https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.1:

> A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request

Please, consider removing `throw new TypeError('Request with GET/HEAD method cannot have body');` from the library, or at least allow the user to bypass this limitation.

**Comment by jimmywarting** [comment]
The browser don't allow you to make any request with a body using a GET.
Browser also have some beef with regards about caching and that no request with a payload should not be cacheable

So b/c browsers have such a big roll in what most request are being made from then then i think it's also bad to enable it and encourage other developer to not rely on a GET request to have any payload. we also have a spec to follow that says that body with a get request should throw

elasticsearch knows this so it also allow you to make a POST request to search for things
IMO i think there is a better http METHODS that they tried to define earlier called [SEARCH](https://www.ietf.org/archive/id/draft-snell-search-method-02.html)
Another way could be to include the query in searchParams to allow for some sort of payload in the url or using some override x-method override with either some header or url param`POST URL?_x-method=GET`

https://stackoverflow.com/a/983458/1008999
https://github.com/node-fetch/node-fetch/issues/1024

**Comment by LinusU** [comment]
This library is an implementation of `fetch`. The specification of `fetch` says to throw an error if there is a body specified in HEAD or GET request:

> 34. If either init["[body](https://fetch.spec.whatwg.org/#dom-requestinit-body)"] [exists](https://infra.spec.whatwg.org/#map-exists) and is non-null or inputBody is non-null, and request’s [method](https://fetch.spec.whatwg.org/#concept-request-method) is `GET` or `HEAD`, then [throw](https://webidl.spec.whatwg.org/#dfn-throw) a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).

ref: https://fetch.spec.whatwg.org/#request-class

Because of this, I think that this feature request is better targeted towards the fetch spec rather than to this implementation?
